### PR TITLE
feat: Add ReactAxisLabelGantt.ts

### PR DESCRIPTION
**Pull request by jbrundage**
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit message includes a "fixes" reference if appropriate.
  - [ ] The commit is signed.
- [ ] The change has been fully tested:
  - [ ] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

**Comment by jbrundage** [comment]
@GordonSmith this is ready for review. The dataflow_timeline.html file.

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 1]
```diff
@@ -0,0 +1,229 @@
+import { HTMLWidget } from "@hpcc-js/common";
+import { Border2 } from "@hpcc-js/layout";
+import { IAxisGanttData } from "./ReactAxisGantt";
+import { ReactTimelineSeries } from "./ReactTimelineSeries";
+import { React, render, TextList } from "@hpcc-js/react";
+
+export class ReactAxisLabelTimeline extends Border2 {
+
+    private _listRenderer: React.FunctionComponent = TextList;
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:9]: No need for the type (will be more accurate without it)

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                rowHeight: this.bucketHeight() + this.gutter(),
+                fontFamily: this.labelFontFamily(),
+                fontSize: this.labelFontSize(),
+                data: this._gantt.data().map(n => n[0]),
+                onItemClick: (evt) => {
+                    console.log("evt", evt);
+                }
+            },
+            lhsNode
+        );
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:60]: This should be moved into a new Widget:  `class TextListWidget extends HTMLWidget`

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        this._centerWA.resize({
+            width: w - _labelWidth,
+            height: contentHeight
+        });
+        this.left().render();
+        const lhsNode = this._leftWA.element().node();
+        lhsNode.style.overflowX = "hidden";
+        lhsNode.style.text = "hidden";
+        lhsNode.style.whiteSpace = "nowrap";
+        lhsNode.style.textOverflow = "ellipsis";
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:88]: These should be published properties of the _leftWA Widget?

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                const scrollTop = node.scrollTop;
+                lhsNode.scrollTop = scrollTop;
+            };
+        });
+    }
+
+    columns(): string[];
+    columns(_: string[]): this;
+    columns(_?: string[]): this | string[] {
+        const retVal = super.columns.apply(this, arguments);
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:101]: `super.columns(_)` should suffice.

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        if (arguments.length > 0) {
+            this._gantt.columns(_);
+        }
+        return retVal;
+    }
+
+    data(): IAxisGanttData[];
+    data(_: IAxisGanttData[]): this;
+    data(_?: IAxisGanttData[]): this | IAxisGanttData[] {
+        const retVal = super.data.apply(this, arguments);
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:111]: `super. data(_)` should suffice.

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        if (arguments.length > 0) {
+            const ganttData: any[] = this.data().map(n => {
+                const ret = [...n];
+                ret[1] = isNaN(n[1] as any) ? new Date(n[1]).getTime() : Number(n[1]);
+                ret[2] = isNaN(n[2] as any) ? new Date(n[2]).getTime() : Number(n[2]);
+                return ret;
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:117]: WUs can span midnight

File: packages/timeline/src/ReactAxisLabelTimeline.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            this._gantt.data(ganttData);
+        }
+        return retVal;
+    }
+
+    resize(_size?: { width: number, height: number }) {
+        const retVal = super.resize.apply(this, arguments);
+
+        return retVal;
+    }
```
Comment (GordonSmith) [inline packages/timeline/src/ReactAxisLabelTimeline.ts:129]: Not needed

File: packages/timeline/src/ReactAxisLabelTimeline.ts (hunk 3 above)
Comment (jbrundage) [reply packages/timeline/src/ReactAxisLabelTimeline.ts:88]: The _leftWA Widget doesn't extend Widget and has no publish properties.

File: packages/timeline/src/ReactAxisLabelTimeline.ts (hunk 6 above)
Comment (jbrundage) [reply packages/timeline/src/ReactAxisLabelTimeline.ts:117]: This implementation supports that

File: packages/layout/src/Border2.ts [hunk 8]
```diff
@@ -187,6 +187,16 @@ export class Border2 extends HTMLWidget {
             .style("width", `${this.width()}px`)
             .style("height", `${this.height()}px`)
             ;
+        
+        const centerNode = this._centerWA.element().node();
+        const context = this;
+        centerNode.onscroll = function () {
+            context.centerScroll(centerNode);
+        };
```
[the remark this answers is not in the document]
Comment (jbrundage) [reply packages/layout/src/Border2.ts:195]: The WidgetDiv class doesn't have a d3 element associated with it and it has no other event handlers (aside from resize).

**Review by GordonSmith — COMMENTED** [review]
I still need to pull local and retest...

**Pull request closed without merging**